src/main/scala/ai/privado/languageEngine/java/tagger/source/SourceTagger.scala

```diff
@@ -0,0 +1,325 @@
+      .foreach(impactedObject => {
+        if (impactedObject.tag.nameExact(Constants.id).l.isEmpty) {
+          storeForTag(builder, impactedObject, ruleCache)(
+            InternalTag.OBJECT_OF_SENSITIVE_CLASS_BY_MEMBER_NAME.toString,
+            taggerPart.ruleInfo.id
+          )
+          storeForTag(builder, impactedObject, ruleCache)(
+            Constants.id,
+            Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME
+          )
+          storeForTag(builder, impactedObject, ruleCache)(Constants.catLevelOne, CatLevelOne.DERIVED_SOURCES.name)
+        }
+        storeForTag(builder, impactedObject, ruleCache)(
+          Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          taggerPart.ruleInfo.id
+        )
+        // Tag for storing memberName in derived Objects -> user --> (email, password)
+        storeForTag(builder, impactedObject, ruleCache)(
+          taggerPart.ruleInfo.id + Constants.underScore + Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          typeDeclMemberName
+        )
+      })
+
+    // To Mark all field Access and getters
+    cpgNodeCache.tagAllFieldAccessAndGetters(builder, typeDeclVal, taggerPart.ruleInfo, typeDeclMemberName, true)
+  }
+}
+
+case class ExtendingDerivedSourceTaggerPart(typeDecl: TypeDecl, memberType: SecondLevelDerivedSourceTaggerPart)
+case class SecondLevelDerivedSourceTaggerPart(typeDecl: TypeDecl, ruleInfo: RuleInfo)
+case class DerivedSourceTaggerPart(typeDecl: TypeDecl, member: Member, ruleInfo: RuleInfo)
+case class CPGNodeCacheForSourceTagger(cpg: Cpg, ruleCache: RuleCache) {
+  val cachedIdentifier: List[Identifier] = cpg.identifier
+    .filterNot(item => item.name.equals(item.name.toUpperCase))
+    .filterNot(_.code.equals(Constants.thisConstant))
+    .l
+  val cachedParameter: List[MethodParameterIn] = cpg.parameter
+    .filterNot(item => item.name.equals(item.name.toUpperCase))
+    .filterNot(_.code.equals(Constants.thisConstant))
+    .l
+  val cachedMember = cpg.member.filterNot(item => item.name.equals(item.name.toUpperCase)).l
+  val cachedFieldAccess = cpg.method
+    .fullNameExact(Operators.fieldAccess, Operators.indirectFieldAccess)
+    .callIn
+    .dedup
+    .l
+  val derivedSourceTaggerParts = ruleCache.getRule.sources
+    .flatMap(rule => {
+      cachedMember
+        .name(rule.combinedRulePattern)
+        .map(memberNode => DerivedSourceTaggerPart(memberNode.typeDecl, memberNode, rule))
+        .l
+    })
+    .toArray
+  val secondLevelDerivedSourceTaggerParts = ruleCache.getRule.sources
+    .flatMap(rule => {
+      cachedMember
+        .name(rule.combinedRulePattern)
+        .map(_.typeDecl)
+        .distinctBy(_.fullName)
+        .map(memberTypeDecl => SecondLevelDerivedSourceTaggerPart(memberTypeDecl, rule))
```

Comment:
Similar optimization as above can be achieved via tagging

Reply:
refer to my above comment